server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -288,8 +288,16 @@ public final void addDynamicMapper(Mapper mapper) {
         // note that existing fields can also receive dynamic mapping updates (e.g. constant_keyword to fix the value)
         if (mappingLookup.getMapper(mapper.name()) == null
             && mappingLookup.objectMappers().containsKey(mapper.name()) == false
-            && newFieldsSeen.add(mapper.name())) {
-            mappingLookup.checkFieldLimit(indexSettings().getMappingTotalFieldsLimit(), newFieldsSeen.size());
+            && dynamicMappers.containsKey(mapper.name()) == false) {
```

Comment:
This may lead to under-counting when we're adding another mapper for the same field that has more sub-fields/multi-fields than the first one. That's probably niche enough, though. We could remove the condition at the expense of potentially over-counting then.